src/scheduler/resv_info.c

```diff
@@ -1636,92 +1656,148 @@ confirm_reservation(status *policy, int pbs_sd, resource_resv *unconf_resv, serv
+
+		if (ns->end_of_chunk)
+			break;
+	}
+
+	return ret;
+}
+
 /**
  * @brief
- * 		Checks the state of all vnodes associated to a reservation and counts the
- * 		number that are unavailable.  Any down node has their nspec->ninfo pointer cleared.
+ * 		Checks the state of all vnodes associated to a reservation and reports
+ * 		if there are any unavailable.  The ninfo ptr of the nspec is cleared for
+ * 		unavailable vnodes so they can be left out when creating the execselect
  *
  * @param[in]	resv	-	resv to check
- * @param[out]	tot_vnodes	-	the total number of vnodes associated to the reservation
- * @param[out]	names_of_down_vnodes	- the string of vnodes that are down that are then
- * 						printed into the log. This has to be of maximum length MAXVNODELIST
  *
- * @return	the number of vnodes down. The total number of vnodes is passed back
- * 			by reference.
+ * @return	1	: there is an unavailable vnode
+ * @retval	0	: all nodes are up and happy
  * @retval	-1	: there is a running job on one of the unavailable nodes.
- * @retval	-2	: error
+ * @retval	-2	: can't map original chunks to resv_nodes
+ * @retval	-3	: error
  */
 int
-check_vnodes_down(resource_resv *resv, int *tot_vnodes, char *names_of_down_vnodes)
+check_vnodes_unavailable(resource_resv *resv)
 {
-	int nn_length; /* number of characters allowable in log buffer */
-	int vnodes_down = 0; /* number of unavailable vnodes */
-	int j;
-	int tot_chk_cnt;
-	int chk_ind = 0;
-	chunk *chk;
-
-	if (resv == NULL || resv->nspec_arr == NULL || tot_vnodes == NULL)
-		return -2;
-
-	*tot_vnodes = 0;
-	chk = resv->select->chunks[0];
-	/* check chunk in the select spec has the number of chunks of that type we need.
-	 * tot_chnk_cnt is the number of chunks of the previous chunks plus the number of this chunk.
-	 * Once we've reached this number of nodes we know we have to move onto the next chunk.
-	 */
-	tot_chk_cnt = chk->num_chunks;
+	int ret = 0;
+	int i;
+	int has_down_node = 0;
+	int has_superchunk = 0;
+	int num_chunks = 0;
+	nspec **chunks_to_remove = NULL;
+	int del_i = 0;
+
+	if (resv == NULL || resv->nspec_arr == NULL)
+		return -3;
+
+	for (i = 0; resv->orig_nspec_arr[i] != NULL; i++) {
+		if (!resv->orig_nspec_arr[i]->end_of_chunk)
+			has_superchunk = 1;
+		num_chunks++;
+	}
 
-	for (j = 0; resv->nspec_arr[j] != NULL; j++) {
-		node_info *ninfo = resv->nspec_arr[j]->ninfo;
-		(*tot_vnodes)++;
+	if (has_superchunk) {
+		if ((chunks_to_remove = malloc((num_chunks + 1) * sizeof(nspec *))) == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			return -3;
+		}
+	}
 
-		if (j == tot_chk_cnt) {
-			chk_ind++;
-			chk = resv->select->chunks[chk_ind];
-			tot_chk_cnt += chk->num_chunks;
+	for (i = 0; resv->orig_nspec_arr[i] != NULL; i++) {
+
+		/* If the original select chunks haven't been mapped to the resv_nodes and the reservation is running, we can't continue */
+		if (resv->resv->resv_state == RESV_RUNNING && resv->orig_nspec_arr[i]->chk == NULL) {
+			free(chunks_to_remove);
```

Comment:
Can we log a message here?

Reply:
We could, but it could get messy.  Any reservation in pre-upgrade to 2020 will not be able to map.  Also, we can't request the parts of a superchunk, so any running reservation can only be reconfirmed once.  After that, the exec_vnode will have more chunks than the select.  This is why I didn't add a message.  I can add one if you think we should.

Reply:
No, it's ok, you can leave it as it is.